### PR TITLE
Refactor `G3WObject`, `Panel` and `Component` classes

### PR DESCRIPTION
**Pull request by Raruto**
## Motivation

1. Make it easier for other plugin developers to understand how internal components work.
2. Start addressing https://github.com/g3w-suite/g3w-client/issues/547
3. Allow plugins to use ES6 classes (brings https://github.com/g3w-suite/g3w-client/pull/574/commits/dff014de180c9b54cf77b856834e27e0dfc038f9 from https://github.com/g3w-suite/g3w-client/pull/451)

## Abstract

- move class definition from `gui/utils/barstack.js` into `core/g3w-barstack.js`
- move class definition from `gui/component/component.js` into `core/g3w-component.js`
- move class definition from `gui/panel.js` into `core/g3w-panel.js`
- new file `component/g3w-catalog.js`
- new file `component/g3w-changemapmenu.js`
- new file `component/g3w-contentsviewer.js`
- new file `component/g3w-form.js`
- new file `component/g3w-map.js`
- new file `component/g3w-metadata.js`
- new file `component/g3w-print.js`
- new file `component/g3w-projectsmenu.js`
- new file `component/g3w-querybuilder.js`
- new file `component/g3w-queryresults.js`
- new file `component/g3w-relationspage.js`
- new file `component/g3w-search.js`
- new file `component/g3w-spatialbookmarks.js`
- new file `component/g3w-tools.js`
- new file `component/g3w-wms.js`
- deprecate setting `search_endpoint = 'ows'` (until removal, the new default is: `'api'`)

Related to: https://github.com/g3w-suite/g3w-client/issues/193

### CODE REVIEW

**Comment by Raruto** [comment]
@volterra79 please avoid rewriting `g3w-*` components too much (eg. https://github.com/g3w-suite/g3w-client/pull/574/commits/d40f577c87ad4798a943fa03c2dc7f81942858ea, https://github.com/g3w-suite/g3w-client/pull/574/commits/ca6c0112071593d93afdc55ff60db067d8b8140b), they don't look nice but there are some reasons why are written/proposed in a such a way..

**Comment by Raruto** [comment]
@volterra79 🥲 sometimes using more specific descriptions would help others analyze/find your edits. 

image

File: src/components/g3w-tools.js [hunk 1]
```diff
@@ -85,13 +85,13 @@ export default function(opts = {}) {
...
-    <div :id="g.name + '-tools'" class="tool-box"><g3w-tool v-for="t in g.tools" :key="t.name" :tool="t" /></div>
-  </li>
-</ul>`,
+        <ul class="g3w-tools treeview-menu">
+          <bar-loader :loading="state.loading"/>
+          <li v-for="g in state.toolsGroups" :key="g.name">
+            <div class="tool-header"><i :class="g3wtemplate.getFontClass('tool')"></i><span>{{ g.name }}</span></div>
+            <div :id="g.name + '-tools'" class="tool-box"><g3w-tool v-for="t in g.tools" :key="t.name" :tool="t" /></div>
+          </li>
+        </ul>`,
```
Comment (Raruto) [inline src/components/g3w-tools.js:94]: @volterra79 in that way you are introducing a lot of whitespaces into the compiled bundle

**Comment by Raruto** [comment]
Needs investagion:

## KO: [object Object]

image

## OK: https://dev.g3wsuite.it/en/map/demo-37/

image

File: src/components/SearchPanel.vue [hunk 2]
```diff
@@ -4,112 +4,558 @@
...
+            <label :for="input.id" class="search-label">
+              <span>{{ input.label || input.attribute }}</span>
+              <span class="skin-color">{{ getLabelOperator(input.operator)}}</span>
+            </label>
+            <div class="input-group date">
+              <input :id="input.id" type='text' class="form-control" />
+              <span class="input-group-addon skin-color">
+                <span :class="g3wtemplate.getFontClass(input.options.format.time ? 'time': 'calendar')"></span>
+              </span>
+            </div>
```
Comment (volterra79) [inline src/components/SearchPanel.vue]: @Raruto to solve issue on datetimepiker click on calendar icon need to move id to div as follow
```suggestion
            <div :id="input.id" class="input-group date">
              <input type='text' class="form-control" />
              <span class="input-group-addon skin-color">
                <span :class="g3wtemplate.getFontClass(input.options.format.time ? 'time': 'calendar')"></span>
              </span>
            </div>
```

File: src/components/SearchPanel.vue [hunk 3]
```diff
@@ -4,112 +4,552 @@
...
+          // parent has invalid value (eg. ALL_VALUE) → set all values to subscribe 
+          if (is_invalid(value) && !is_autocomplete(s)) {
+            setTimeout(() => s.values = [...s._values]);
+          }
+
+          s.value = 'selectfield' === s.type ? SEARCH_ALLVALUE : null;
+        });
+
+        // value is empty → disable dependants inputs
+        if (!value || value === SEARCH_ALLVALUE) {
```
Comment (Raruto) [inline src/components/SearchPanel.vue]: is this correct?

```suggestion
        if (is_invalid(value)) {
```

ie. the original version also excludes zero

**Review by volterra79 — APPROVED** [review]

File: src/components/SearchPanel.vue (hunk 2 above)
Comment (Raruto) [reply src/components/SearchPanel.vue]: @volterra79 Your suggestion triggers a console.warning.

what do you think about label element? ref: https://github.com/g3w-suite/g3w-client/pull/574/commits/397fda59f94d717d19c149ac682f3df88cab09b5

Comment (volterra79) [reply src/components/SearchPanel.vue]: > @volterra79 Your suggestion triggers a console.warning.
> 
> what do you think about label element? ref: [397fda5](https://github.com/g3w-suite/g3w-client/commit/397fda59f94d717d19c149ac682f3df88cab09b5)

reference documentation https://getdatepicker.com/4/ suggest to set id on div

**Pull request merged**